lib/ex_doc/formatter/markdown.ex

```diff
@@ -0,0 +1,539 @@
+    base = [
+      apps: config.apps,
+      deps: config.deps,
+      ext: ext,
+      extras: extra_paths(config),
+      skip_undefined_reference_warnings_on: config.skip_undefined_reference_warnings_on,
+      skip_code_autolink_to: config.skip_code_autolink_to,
+      filtered_modules: filtered_modules
+    ]
+
+    project_nodes
+    |> Task.async_stream(
+      fn node ->
+        language = node.language
+
+        autolink_opts =
+          [
+            current_module: node.module,
+            file: node.moduledoc_file,
+            line: node.moduledoc_line,
+            module_id: node.id,
+            language: language
+          ] ++ base
+
+        docs =
+          for child_node <- node.docs do
+            id = id(node, child_node)
+
+            autolink_opts =
+              autolink_opts ++
+                [
+                  id: id,
+                  line: child_node.doc_line,
+                  file: child_node.doc_file,
+                  current_kfa: {child_node.type, child_node.name, child_node.arity}
+                ]
+
+            specs = Enum.map(child_node.specs, &language.format_spec(&1))
+            child_node = %{child_node | specs: specs}
+            render_doc(child_node, language, autolink_opts, opts, 4)
+          end
+
+        %{
+          render_doc(node, language, [{:id, node.id} | autolink_opts], opts, 2)
+          | docs: docs
+        }
+      end,
+      timeout: :infinity
+    )
+    |> Enum.map(&elem(&1, 1))
+  end
+
+  defp render_doc(%{doc: nil} = node, _language, _autolink_opts, _opts, _base_heading),
+    do: node
+
+  defp render_doc(
+         %{doc: _doc, source_doc: source_doc} = node,
+         _language,
+         _autolink_opts,
+         _opts,
+         base_heading
+       ) do
+    # rendered = autolink_and_render(doc, language, autolink_opts, opts)
+    rendered = rewrite_headings(source_doc["en"], base_heading)
+    %{node | rendered_doc: rendered}
+  end
+
+  defp rewrite_headings(markdown, base_heading)
+       when is_binary(markdown) and is_integer(base_heading) and base_heading >= 1 do
+    {:ok, document} = MDEx.parse_document(markdown)
+
+    document =
+      case find_lowest_heading(document) do
+        lowest_heading when lowest_heading >= base_heading ->
+          document
+
+        lowest_heading ->
+          levels_to_bump = base_heading - lowest_heading
+
+          bump_levels(document, levels_to_bump)
+      end
+
+    MDEx.to_commonmark!(document)
+  end
+
+  defp find_lowest_heading(document) when is_struct(document, MDEx.Document) do
+    Enum.reduce_while(document, 6, fn
+      %MDEx.Heading{level: 1}, _lowest_level ->
+        {:halt, 1}
+
+      %MDEx.Heading{level: level}, lowest_level when level < lowest_level ->
+        {:cont, level}
+
+      _, lowest_level ->
+        {:cont, lowest_level}
+    end)
+  end
+
+  defp bump_levels(%MDEx.Document{nodes: nodes} = document, levels_to_bump) do
```

Comment:
Hi @eksperimental you could leverage the Access protocol if you prefer. Here's an example: https://github.com/leandrocp/mdex/commit/d44bdd69e3ab6fa506a93a46de102b4c1fffbdfb

Reply:
Thank you. Applied in 805b8ba